README.md
# LifeOmic App Tools

[![Build
Status](https://travis-ci.org/lifeomic/app-tools.svg?branch=master)](https://travis-ci.org/lifeomic/app-tools)

Provides a set of utilities for developing custom web apps against the LifeOmic PHC API.

## Installation

```bash
yarn install @lifeomic/app-tools
```

## Auth

```javascript
import { LOAuth } from '@lifeomic/app-tools';

// Setup
const appAuth = new LOAuth({
  clientId: '<clientId>',
  authorizationUri: 'https://lifeomic-prod-us.auth.us-east-2.amazoncognito.com/oauth2/authorize',
  accessTokenUri: 'https://lifeomic-prod-us.auth.us-east-2.amazoncognito.com/oauth2/token',
  redirectUri: 'http://localhost:3000/callback',
  logoutUri: 'https://lifeomic-prod-us.auth.us-east-2.amazoncognito.com/logout',
  logoutRedirectUri: 'http://localhost:3000/logout',
  scopes: ['openid']
});

appAuth.startAutomaticTokenRefresh().then(() => {
  const account = '<myaccountid>';
  const resourceType = 'Patient';
  const project = '<myprojectId>';

  // Sign adds access_token etc. to your request options
  const request = await appAuth.sign({
    method: 'GET',
    url: `https://fhir.us.lifeomic.com/${account}/dstu3/${resourceType}?_tag=http%3A%2F%2Flifeomic.com%2Ffhir%2Fdataset%${project}&pageSize=5`
  });

  const response = await fetch(request.url, request);
  console.log(response);
});
```

### Token Storage

By default `LOAuth` will store the token in local storage and hydrate any existing session from there. To overwrite the storage location you can pass a storage object in options along with a custom storage key.

```ts
import { LOAuth } from '@lifeomic/app-tools';

const appAuth = new LOAuth({
  ...oauthConfig,
  storageKey: 'my-super-awesome-key',
  storage: {
    getItem(key: string) {
      return sessionStorage.getItem(key);
    },
    setItem(key: string, value: string) {
      return sessionStorage.setItem(key, value);
    },
    removeItem(key: string) {
      return sessionStorage.removeItem(key)
    }
  }
});

// or simply
const appAuth = new LOAuth({
  ...oauthConfig,
  storageKey: 'my-super-awesome-key',
  storage: sessionStorage,
});
```

To opt-out of storing the token, you can pass in noop functions for storage

```ts
const noop = () => null;

const appAuth = new LOAuth({
  ...oauthConfig,
  storageKey: 'my-super-awesome-key',
  storage: {
    getItem: noop,
    setItem: noop,
    removeItem: noop
  }
});
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [2.0.2] - 2019-12-31
### Changed
- Bumped client-oauth2 library to include less strict/unnecessary validation on
redirectUri + post-auth pathname matching. This allows for intermediary
redirects in the authorize flow.

## [2.0.0] - 2019-12-26
### Breaking
- Switch to base64 encoding to improve state query param parsing consistency
  [#13](https://github.com/lifeomic/app-tools/pull/13). This is unlikely to
  affect most consumers.


test/LOAuth.test.ts
const queryString = require('query-string');
const ClientOAuth2 = require('client-oauth2');
const globals = require('../src/globals');
import LOAuth from '../src/LOAuth';

jest.mock('client-oauth2', () => jest.fn().mockImplementation((...params) => {
  ClientOAuth2Ctor(...params);
  return clientOAuth2;
}));

jest.mock('../src/globals', () => ({
  window: {
    location: {
      href: 'set-in-before-each!',
      search: 'set-in-before-each!',
      protocol: 'https:',
      hostname: 'unit-test',
      pathname: ''
    },
    history: {
      replaceState: jest.fn()
    },
    document: {
      title: 'title'
    },
    localStorage: {
      getItem: jest.fn(() => null),
      setItem: jest.fn(),
      removeItem: jest.fn()
    },
    setInterval: jest.fn(),
    clearInterval: jest.fn(),
    fetch: jest.fn().mockResolvedValue({ json: async () => tokenResponseJson, ok: true })
  }
}));

let mockToken, tokenResponseJson, ctorParams, auth;

const clientOAuth2 = {
  code: {
    getToken: jest.fn(),
    getUri: jest.fn()
  },
  createToken: jest.fn()
};
const ClientOAuth2Ctor = jest.fn();

beforeEach(() => {
  globals.window.location = {};
  tokenResponseJson = {
    access_token: 'access_token',
    expires_in: 3600,
    id_token: 'id_token',
    refresh_token: 'refresh_token',
    token_type: 'Bearer'
  };
  mockToken = {
    refreshToken: 'refresh-token',
    sign: jest.fn(),
    expiresIn: jest.fn(),
    expires: {
      getTime: jest.fn(() => 100)
    },
    data: {
      ...tokenResponseJson
    }
  };
  clientOAuth2.code.getToken.mockResolvedValue(mockToken);
  clientOAuth2.code.getUri.mockReturnValue('http://integ-test/oauth-uri');
  clientOAuth2.createToken.mockReturnValue(mockToken);
  globals.window.localStorage.getItem.mockReturnValue(null);
  ctorParams = {
    clientId: '7cbji8hkta84ons79j34qcfdci',
    authorizationUri: 'https://integ-test/authorize',
    accessTokenUri: 'https://integ-test/token',
    logoutUri: 'https://integ-test/logout',
    redirectUri: 'http://localhost:3000/callback',
    logoutRedirectUri: 'http://localhost:3000/logout',
    scopes: ['openid']
  };

  Object.assign(globals.window.location, {
    href: 'https://unit-test?client_id=foo&authorization_code=bar',
    search: '?client_id=foo&authorization_code=bar'
  });
});

const ctorRequired = [
  'clientId',
  'authorizationUri',
  'accessTokenUri',
  'redirectUri',
  'logoutUri',
  'logoutRedirectUri',
  'scopes'
];

for (const param of ctorRequired) {
  test(`ctor throws if param ${param} is missing`, async () => {
    // eslint-disable-next-line security/detect-object-injection
    delete ctorParams[param];
    expect(() => new LOAuth(ctorParams)).toThrow(new RegExp(param));
  });
}

test('ctor captures LO query string parameters for application state', async () => {
  globals.window.location.href = 'https://unit-test?account=myaccount&projectId=myproject';
  globals.window.location.search = '?account=myaccount&projectId=myproject';
  const expectedState = btoa(JSON.stringify({ account: 'myaccount', projectId: 'myproject' }));
  auth = new LOAuth(ctorParams);
  expect(ClientOAuth2Ctor.mock.calls[0][0].state).toEqual(expectedState);
});

test('ctor captures application state from OAuth state param', async () => {
  ctorParams.appState = {
    account: 'myaccount', // This will override queryString params (account lock down for app)
    myField: 'myValue'
  };
  const expectedState = btoa(JSON.stringify(ctorParams.appState));
  globals.window.location.href = `https://unit-test?account=otheraccount`;
  auth = new LOAuth(ctorParams);
  expect(ClientOAuth2Ctor).toBeCalledTimes(1);
  expect(ClientOAuth2Ctor.mock.calls[0][0].state).toEqual(expectedState);
});

test('ctor passes empty state when appropriate', async () => {
  delete ctorParams.appState;
  const expectedState = undefined;
  globals.window.location.href = 'https://unit-test';
  globals.window.location.search = '';
  auth = new LOAuth(ctorParams);
  expect(ClientOAuth2Ctor.mock.calls[0][0].state).toEqual(expectedState);
});

test('ctor encodes pathname in state object', async () => {
  const pathname = '/deep/client/path';
  const expectedState = btoa(JSON.stringify({ pathname }));
  globals.window.location.pathname = pathname;
  auth = new LOAuth(ctorParams);
  expect(ClientOAuth2Ctor.mock.calls[0][0].state).toEqual(expectedState);
});

describe('with auth successfully created', () => {
  beforeEach(() => {
    auth = new LOAuth(ctorParams);
  });

  test('refreshAccessToken attempts to exchange authorization code for access token', async () => {
    globals.window.location.protocol = 'https:';
    globals.window.location.hostname = 'unit-test';
    globals.window.location.pathname = '/test/deep/links/';
    auth = new LOAuth(ctorParams);
    
    await auth.refreshAccessToken();

    const client = ClientOAuth2.mock.results[0].value;
    expect(client.code.getToken).toBeCalledTimes(1);
    expect(client.code.getToken.mock.calls[0][0]).toBe(
      'https://unit-test?client_id=foo&authorization_code=bar'
    );
    expect(globals.window.history.replaceState).toBeCalledTimes(1);
    expect(globals.window.history.replaceState.mock.calls[0][2]).toBe(
      'https://unit-test/test/deep/links/'
    );
    expect(globals.window.localStorage.setItem.mock.calls).toMatchSnapshot();
  });

  test('refreshAccessToken decodes and exposes appState', async () => {
    const appState = { account: 'myaccount', projectId: 'myproject', customAppField: 'val', pathname: '/client/deep/link/url/' };
    const encodedState = btoa(JSON.stringify(appState));
    globals.window.location.search = `?client_id=foo&authorization_code=bar&state=${encodedState}`;
    globals.window.location.href = `https://unit-test${globals.window.location.search}`;
    await auth.refreshAccessToken();

    expect(auth.appState.account).toBe('myaccount');
    expect(auth.appState.projectId).toBe('myproject');
    expect(auth.appState.customAppField).toBe('val');
    expect(auth.appState.pathname).toBe('/client/deep/link/url/');
  });

  test('refreshAccessToken does nothing if token already captured', async () => {
    await auth.refreshAccessToken(); // token captured
    await auth.refreshAccessToken(); // no-op

    expect(
      ClientOAuth2.mock.results[0].value.code.getToken.mock.calls.length
    ).toBe(1);
    expect(globals.window.fetch).toBeCalledTimes(0);
  });

  test('refreshAccessToken performs refresh_token if called with expiringRefresh', async () => {
    await auth.refreshAccessToken(); // token captured
    await auth.refreshAccessToken({ expiringRefresh: true });

    expect(globals.window.fetch).toBeCalledTimes(1);
    expect(globals.window.fetch.mock.calls[0][0]).toBe(ctorParams.accessTokenUri);
    expect(globals.window.fetch.mock.calls[0][1].body).toEqual(
      queryString.stringify({
        client_id: ctorParams.clientId,
        grant_type: 'refresh_token',
        refresh_token: mockToken.refreshToken,
        redirect_uri: ctorParams.redirectUri
      })
    );
    expect(ClientOAuth2.mock.results[0].value.createToken).toBeCalledTimes(1);
    expect(globals.window.localStorage.setItem.mock.calls).toMatchSnapshot();
  });

  test('refreshAccessToken expiringRefresh handles refresh error', async () => {
    await auth.refreshAccessToken(); // token captured

    expect(globals.window.localStorage.setItem).toBeCalledTimes(1);

    globals.window.fetch.mockReturnValueOnce({
      ok: false,
      status: 400,
      error: 'invalid_grant'
    });

    await auth.refreshAccessToken({ expiringRefresh: true });

    expect(globals.window.fetch).toBeCalledTimes(1);
    expect(globals.window.fetch.mock.calls[0][0]).toBe(ctorParams.accessTokenUri);
    expect(globals.window.fetch.mock.calls[0][1].body).toEqual(
      queryString.stringify({
        client_id: ctorParams.clientId,
        grant_type: 'refresh_token',
        refresh_token: mockToken.refreshToken,
        redirect_uri: ctorParams.redirectUri
      })
    );
    expect(ClientOAuth2.mock.results[0].value.createToken).toBeCalledTimes(0);
    expect(globals.window.localStorage.setItem).toBeCalledTimes(2);

    expect(globals.window.localStorage.setItem.mock.calls).toMatchSnapshot();
  });

  test('refreshAccessToken redirects to login upon error if no token is in storage', async () => {
    globals.window.location.href = 'https://unit-test'; // No client_id etc.
    clientOAuth2.code.getToken.mockRejectedValue(new Error('unit test'));
    await auth.refreshAccessToken();

    const loginUri = await ClientOAuth2.mock.results[0].value.code.getUri();

    expect(globals.window.location.href).toBe(loginUri);
  });

  test('refreshAccessToken uses token from storage upon error', async () => {
    const appUrl = 'https://unit-test';
    globals.window.location.href = appUrl;
    clientOAuth2.code.getToken.mockRejectedValue(new Error('unit test'));
    clientOAuth2.createToken.mockReturnValue({
      ...mockToken,
      expires: {
        getTime: () => Date.now() * 2
      }
    });
    globals.window.localStorage.getItem.mockReturnValueOnce(
      JSON.stringify({
        ...tokenResponseJson,
        expires: 100
      })
    );
    await auth.refreshAccessToken();

    expect(globals.window.location.href).toBe(appUrl);
    expect(globals.window.fetch).not.toBeCalled();
  });

  test('refreshAccessToken refresh token from storage upon error if it is expired', async () => {
    globals.window.location.href = 'https://unit-test'; // No client_id etc.
    clientOAuth2.code.getToken.mockRejectedValue(new Error('unit test'));
    globals.window.localStorage.getItem.mockReturnValueOnce(
      JSON.stringify({
        ...tokenResponseJson,
        expires: 0
      })
    );

    await auth.refreshAccessToken();
    expect(globals.window.fetch).toBeCalled();
  });

  test('LOAuth uses storage passed to it', async () => {
    const mockStorageKey = 'mock-storage-key';
    const mockStorage: LOAuth.Storage = {
      getItem: jest.fn(() => {
        JSON.stringify({
          ...tokenResponseJson,
          expires: 0
        });
      }),
      setItem: jest.fn(),
      removeItem: jest.fn()
    };
    const mockAuth = new LOAuth({
      ...ctorParams,
      storageKey: mockStorageKey,
      storage: mockStorage
    });

    globals.window.location.href = 'https://unit-test'; // No client_id etc.
    clientOAuth2.code.getToken.mockRejectedValueOnce(new Error('unit test'));

    await mockAuth.refreshAccessToken();
    await mockAuth.refreshAccessToken({ expiringRefresh: true });
    await mockAuth.logout();
    expect(mockStorage.setItem).toBeCalledWith(
      mockStorageKey,
      expect.any(String)
    );
    expect(mockStorage.getItem).toBeCalledWith(mockStorageKey);
    expect(mockStorage.removeItem).toBeCalledWith(mockStorageKey);
  });

  test('refreshAccessToken does nothing if token already captured', async () => {
    await auth.refreshAccessToken(); // token captured
    await auth.refreshAccessToken(); // no-op

    expect(ClientOAuth2.mock.results[0].value.code.getToken).toBeCalledTimes(1);
    expect(globals.window.fetch).toBeCalledTimes(0);
  });

  test('refreshAccessToken redirects to login upon error', async () => {
    globals.window.location.href = 'https://unit-test'; // No client_id etc.
    clientOAuth2.code.getToken.mockRejectedValue(new Error('unit test'));
    await auth.refreshAccessToken();

    const loginUri = await ClientOAuth2.mock.results[0].value.code.getUri();
    expect(globals.window.location.href).toBe(loginUri);
  });

  const startAutomaticTokenRefreshConfigs = [
    {
      interval: 5,
      refreshWindow: 5 * 1000,
      override: true,
      testTitle: 'startAutomaticTokenRefresh gets token and automatically refreshes before expiration with overrides'
    },
    {
      // Defaults
      interval: 30 * 1000,
      refreshWindow: 5,
      testTitle: 'startAutomaticTokenRefresh gets token and automatically refreshes before expiration with defaults'
    }
  ];
  for (const config of startAutomaticTokenRefreshConfigs) {
    test(config.testTitle, async () => {
      auth.refreshAccessToken = jest.fn();
      const options = config.override ? {
        interval: config.interval,
        refreshWindow: config.refreshWindow
      } : {};
      await auth.startAutomaticTokenRefresh(options);
      mockToken.expires = new Date(Date.now() + ((config.refreshWindow + 2) * 60 * 1000));
      auth.token = mockToken;

      expect(auth.refreshAccessToken).toBeCalledTimes(1) ; // Initial retrieval of token from URL
      expect(globals.window.setInterval).toBeCalledTimes(1);
      expect(globals.window.setInterval.mock.calls[0][1]).toBe(config.interval);
      // Mock setInterval func invocation
      globals.window.setInterval.mock.calls[0][0].bind(auth)();
      expect(auth.refreshAccessToken).toBeCalledTimes(1); // Hasn't expired yet

      mockToken.expires = new Date(Date.now() + ((config.refreshWindow - 1) * 60 * 1000));
      globals.window.setInterval.mock.calls[0][0].bind(auth)();
      expect(auth.refreshAccessToken).toBeCalledTimes(2); // Expired and called
      expect(auth.refreshAccessToken.mock.calls[1][0]).toEqual({ expiringRefresh: true });
    });
  }

  test('stopAutomaticTokenRefresh calls clearInterval if refreshInterval is set', async () => {
    await auth.stopAutomaticTokenRefresh();
    expect(globals.window.clearInterval).toBeCalledTimes(0);

    auth.refreshInterval = () => {};

    await auth.stopAutomaticTokenRefresh();
    expect(globals.window.clearInterval).toBeCalledTimes(1);
    expect(globals.window.clearInterval.mock.calls[0][0]).toBe(auth.refreshInterval);
  });

  test('sign uses client-oauth2 sign if token present', async () => {
    auth.token = mockToken;
    const requestOptions = {
      method: 'GET',
      url: 'https://integ-test/api'
    };
    await auth.sign(requestOptions);
    expect(mockToken.sign).toBeCalledTimes(1);
    expect(mockToken.sign.mock.calls[0][0]).toEqual(requestOptions);
  });

  test('sign throws if not token captured', async () => {
    auth.token = null;
    const requestOptions = {
      method: 'GET',
      url: 'https://integ-test/api'
    };
    expect(auth.sign(requestOptions)).rejects.toThrow();
  });

  test('logout redirects to the logoutUri', async () => {
    auth.stopAutomaticTokenRefresh = jest.fn();
    await auth.logout();

    expect(auth.stopAutomaticTokenRefresh).toBeCalledTimes(1);
    const qs = queryString.stringify({
      client_id: ctorParams.clientId,
      logout_uri: ctorParams.logoutRedirectUri
    });
    expect(globals.window.location.href).toBe(`${ctorParams.logoutUri}?${qs}`);
  });
});


src/__tests__/LOAuth.test.ts
const queryString = require('query-string');
const ClientOAuth2 = require('client-oauth2');
const globals = require('../globals');
import LOAuth from '../LOAuth';

jest.mock('client-oauth2', () => jest.fn().mockImplementation(() => clientOAuth2));

jest.mock('../globals', () => ({
  window: {
    location: {
      href: 'https://unit-test?client_id=foo&authorization_code=bar',
      protocol: 'https:',
      hostname: 'unit-test'
    },
    history: {
      replaceState: jest.fn()
    },
    document: {
      title: 'title'
    },
    setInterval: jest.fn(),
    clearInterval: jest.fn(),
    fetch: jest.fn().mockResolvedValue({ json: async () => tokenResponseJson })
  }
}));

let mockToken,
  tokenResponseJson,
  ctorParams,
  auth;

const clientOAuth2 = {
  code: {
    getToken: jest.fn(),
    getUri: jest.fn()
  },
  createToken: jest.fn()
};

beforeEach(() => {
  mockToken = {
    refreshToken: 'refresh-token',
    sign: jest.fn()
  };
  clientOAuth2.code.getToken.mockResolvedValue(mockToken);
  clientOAuth2.code.getUri.mockResolvedValue('http://integ-test/oauth-uri');
  clientOAuth2.createToken.mockResolvedValue(mockToken);
  Object.assign(global, );
  tokenResponseJson = {
    access_token: 'access_token',
    expires_in: 3600,
    id_token: 'id_token',
    refresh_token: 'refresh_token',
    token_type: 'Bearer'
  };
  ctorParams = {
    clientId: '7cbji8hkta84ons79j34qcfdci',
    authorizationUri: 'https://integ-test/authorize',
    accessTokenUri: 'https://integ-test/token',
    logoutUri: 'https://integ-test/logout',
    redirectUri: 'http://localhost:3000/callback',
    logoutRedirectUri: 'http://localhost:3000/logout',
    scopes: ['openid']
  };

  auth = new LOAuth(ctorParams);
});

const ctorRequired = [
  'clientId',
  'authorizationUri',
  'accessTokenUri',
  'redirectUri',
  'logoutUri',
  'logoutRedirectUri',
  'scopes'
];

for (const param of ctorRequired) {
  test(`ctor throws if param ${param} is missing`, async () => {
    // eslint-disable-next-line security/detect-object-injection
    delete ctorParams[param];
    expect(() => new LOAuth(ctorParams)).toThrow(new RegExp(param));
  });
}

test('refreshAccessToken attempts to exchange authorization code for access token', async () => {
  await auth.refreshAccessToken();

  const client = ClientOAuth2.mock.results[0].value;
  expect(client.code.getToken.mock.calls.length).toBe(1);
  expect(client.code.getToken.mock.calls[0][0]).toBe('https://unit-test?client_id=foo&authorization_code=bar');
  expect(globals.window.history.replaceState.mock.calls.length).toBe(1);
  expect(globals.window.history.replaceState.mock.calls[0][2]).toBe('https://unit-test');
});

test('refreshAccessToken does nothing if token already captured', async () => {
  await auth.refreshAccessToken(); // token captured
  await auth.refreshAccessToken(); // no-op

  expect(ClientOAuth2.mock.results[0].value.code.getToken.mock.calls.length).toBe(1);
  expect(globals.window.fetch.mock.calls.length).toBe(0);
});

test('refreshAccessToken performs refresh_token if called with expiringRefresh', async () => {
  await auth.refreshAccessToken(); // token captured
  await auth.refreshAccessToken({ expiringRefresh: true });

  expect(globals.window.fetch.mock.calls.length).toBe(1);
  expect(globals.window.fetch.mock.calls[0][0]).toBe(ctorParams.accessTokenUri);
  expect(globals.window.fetch.mock.calls[0][1].body).toEqual(queryString.stringify({
    client_id: ctorParams.clientId,
    grant_type: 'refresh_token',
    refresh_token: mockToken.refreshToken,
    redirect_uri: ctorParams.redirectUri
  }));
  expect(ClientOAuth2.mock.results[0].value.createToken.mock.calls.length).toBe(1);
});

test('refreshAccessToken redirects to login upon error', async () => {
  globals.window.location.href = 'https://unit-test'; // No client_id etc.
  clientOAuth2.code.getToken.mockRejectedValue(new Error('unit test'));
  await auth.refreshAccessToken();

  const loginUri = await ClientOAuth2.mock.results[0].value.code.getUri();
  expect(globals.window.location.href).toBe(loginUri);
});

const startAutomaticTokenRefreshConfigs = [
  {
    interval: 5,
    refreshWindow: 5 * 1000,
    override: true,
    testTitle: 'startAutomaticTokenRefresh gets token and automatically refreshes before expiration with overrides'
  },
  {
    // Defaults
    interval: 30 * 1000,
    refreshWindow: 5,
    testTitle: 'startAutomaticTokenRefresh gets token and automatically refreshes before expiration with defaults'
  }
];
for (const config of startAutomaticTokenRefreshConfigs) {
  test(config.testTitle, async () => {
    auth.refreshAccessToken = jest.fn();
    const options = config.override ? {
      interval: config.interval,
      refreshWindow: config.refreshWindow
    } : {};
    await auth.startAutomaticTokenRefresh(options);
    mockToken.expires = new Date(Date.now() + ((config.refreshWindow + 2) * 60 * 1000));
    auth.token = mockToken;

    expect(auth.refreshAccessToken.mock.calls.length).toBe(1); // Initial retrieval of token from URL
    expect(globals.window.setInterval.mock.calls.length).toBe(1);
    expect(globals.window.setInterval.mock.calls[0][1]).toBe(config.interval);
    // Mock setInterval func invocation
    globals.window.setInterval.mock.calls[0][0].bind(auth)();
    expect(auth.refreshAccessToken.mock.calls.length).toBe(1); // Hasn't expired yet

    mockToken.expires = new Date(Date.now() + ((config.refreshWindow - 1) * 60 * 1000));
    globals.window.setInterval.mock.calls[0][0].bind(auth)();
    expect(auth.refreshAccessToken.mock.calls.length).toBe(2); // Expired and called
    expect(auth.refreshAccessToken.mock.calls[1][0]).toEqual({ expiringRefresh: true });
  });
}

test('stopAutomaticTokenRefresh calls clearInterval if refreshInterval is set', async () => {
  await auth.stopAutomaticTokenRefresh();
  expect(globals.window.clearInterval.mock.calls.length).toBe(0);

  auth.refreshInterval = () => {};

  await auth.stopAutomaticTokenRefresh();
  expect(globals.window.clearInterval.mock.calls.length).toBe(1);
  expect(globals.window.clearInterval.mock.calls[0][0]).toBe(auth.refreshInterval);
});

test('sign uses client-oauth2 sign if token present', async () => {
  auth.token = mockToken;
  const requestOptions = {
    method: 'GET',
    url: 'https://integ-test/api'
  };
  await auth.sign(requestOptions);
  expect(mockToken.sign.mock.calls.length).toBe(1);
  expect(mockToken.sign.mock.calls[0][0]).toEqual(requestOptions);
});

test('sign throws if not token captured', async () => {
  auth.token = null;
  const requestOptions = {
    method: 'GET',
    url: 'https://integ-test/api'
  };
  expect(auth.sign(requestOptions)).rejects.toThrow();
});

test('logout redirects to the logoutUri', async () => {
  auth.stopAutomaticTokenRefresh = jest.fn()
  await auth.logout();

  expect(auth.stopAutomaticTokenRefresh.mock.calls.length).toBe(1);
  const qs = queryString.stringify({
    client_id: ctorParams.clientId,
    logout_uri: ctorParams.logoutRedirectUri
  });
  expect(globals.window.location.href).toBe(`${ctorParams.logoutUri}?${qs}`);
});


prePush.sh
# pre-push hook args: <local_ref> <local_sha1> <remote_ref> <remote_sha1>
local_sha1=$1
remote_sha1=$3

# note: this needs to be the same as ignoreCommitPattern in .auto-changelog
# to avoid including CHANGELOG bump commits in the changelog changes
changelog_update_msg='pre-push auto-changelog update'

echo "Pre-push: checking to see if the CHANGELOG should be updated..."

# fetch latest refs
git remote update
num_new_commits_compared_to_master=$(git rev-list HEAD...origin/master --count)

# if there are new commits..
if [ -n $num_new_commits_compared_to_master ]; then
  changelog_already_updated=$(git diff --name-only HEAD~1 HEAD | grep CHANGELOG.md)
  # and the latest commit didn't update the CHANGELOG
  if [ ! $changelog_already_updated ]; then
    yarn auto-changelog --unreleased && git add CHANGELOG.md && git commit -m "$changelog_update_msg"
  fi
fi


src/LOAuth.ts
import * as ClientOAuth2 from 'client-oauth2';
const queryString = require('query-string');
import { window } from './globals';

// Defaults:
const AUTH_TOKEN_INTERVAL = 30 * 1000; // every 30 seconds
const AUTH_REFRESH_WINDOW = 5; // minutes before token expiration that we renew
const AUTH_STORAGE_KEY = 'lo-app-tools-auth';

const LO_QUERY_STRING_PARAMS = [
  'account',
  'projectId',
  'cohortId'
];

class LOAuth {
  private client: ClientOAuth2;
  private clientOptions: LOAuth.Config;
  private token: LOAuth.Token;
  private refreshInterval: number;
  appState: Record<string, string>;

  constructor(options: LOAuth.Config) {
    const required = [
      'clientId',
      'authorizationUri',
      'accessTokenUri',
      'redirectUri',
      'logoutUri',
      'logoutRedirectUri',
      'scopes'
    ];
    for (const param of required) {
      // eslint-disable-next-line security/detect-object-injection
      if (!options[param]) {
        throw new Error(`LOAuth ctor param ${param} is required`);
      }
    }

    const state = this._getStateForClientOAuth(options);

    this.client = new ClientOAuth2({
      clientId: options.clientId,
      authorizationUri: options.authorizationUri,
      accessTokenUri: options.accessTokenUri,
      redirectUri: options.redirectUri,
      scopes: options.scopes,
      state
    });
    this.clientOptions = options;
    this.clientOptions.storageKey = options.storageKey || AUTH_STORAGE_KEY;
    this.clientOptions.storage = options.storage || window.localStorage;
  }

  _getAppUri () {
    this._decodeAppState();
    const protocol = window.location.protocol;
    const hostname = window.location.hostname;
    const port = window.location.port ? ':' + window.location.port : '';
    const queryParameters = new URLSearchParams();
    LO_QUERY_STRING_PARAMS.forEach(param => {
      if (this.appState[param]) {
        queryParameters.set(param, this.appState[param]);
      }
    });
    let queryString = queryParameters.toString();
    queryString = queryString.length ? `?${queryString}` : '';
    const pathname = this.appState.pathname || '';
    return `${protocol}//${hostname}${port}${pathname}${queryString}`;
  }

  _decodeAppState () {
    const queryParameters = new URLSearchParams(window.location.search);
    try {
      this.appState = {};

      // If initial load, read from queryStrings
      LO_QUERY_STRING_PARAMS.forEach(param => {
        if (queryParameters.get(param)) {
          this.appState[param] = queryParameters.get(param);
        }
      });

      // After redirect pathname will be the root directory, but don't add that to
      // the appState object because it messes up the Object key ordering and causes
      // the state comparison to fail
      if (window.location.pathname && window.location.pathname !== '/') {
        this.appState.pathname = window.location.pathname;
      }

      // If after login flow, decode from state
      const queryState = queryParameters.get('state');
      if (queryState) {
        Object.assign(this.appState, JSON.parse(atob(queryState)));
      }
    } catch (error) {
      console.warn(error, 'Error occurred parsing state query string parameter');
    }
  }

  private _getStateForClientOAuth(options: LOAuth.Config) {
    this._decodeAppState();

    const state = {
      ...this.appState,
      ...options.appState
    };

    if (Object.keys(state).length === 0) {
      return undefined;
    }

    return btoa(JSON.stringify(state));
  }

  private _storeTokenData(token: LOAuth.Token) {
    const { storage, storageKey } = this.clientOptions;
    const data: LOAuth.TokenData = {
      ...token.data,
      expires: token.expires.getTime()
    };
    storage.setItem(storageKey, JSON.stringify(data));
  }

  private _getTokenDataFromStorage(): LOAuth.TokenData {
    const { storage, storageKey } = this.clientOptions;
    const value = storage.getItem(storageKey);

    return value ? JSON.parse(value) : null;
  }

  private _removeTokenDataFromStorage() {
    const { storage, storageKey } = this.clientOptions;
    storage.removeItem(storageKey);
  }

  private _isTokenExpiring(options: LOAuth.RefreshOptions = {}) {
    const refreshWindow =
      options.refreshWindow !== undefined
        ? options.refreshWindow
        : AUTH_REFRESH_WINDOW;
    return (
      Math.floor((this.token.expires.getTime() - Date.now()) / 1000 / 60) <=
      refreshWindow
    );
  }

  public async refreshAccessToken(options: LOAuth.RefreshOptions = {}) {
    options = Object.assign({}, this.clientOptions, options);
    try {
      if (!this.token) {
        // Initial auth token exchange
        this.token = await this.client.code.getToken(
          window.location.href,
          options
        );

        this._storeTokenData(this.token);

        // Remove client_id / code from URL
        window.history.replaceState(
          {},
          window.document.title,
          this._getAppUri()
        );
      } else if (options.expiringRefresh) {
        // Token refresh
        const response = await window.fetch(this.clientOptions.accessTokenUri, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded'
          },
          body: queryString.stringify({
            client_id: options.clientId,
            grant_type: 'refresh_token',
            refresh_token: this.token.refreshToken,
            redirect_uri: this.clientOptions.redirectUri
          })
        });

        // Only create/store tokens from a valid response
        if (response.ok) {
          const responseJson = await response.json();

          this.token = this.client.createToken(
            Object.assign(
              {
                refresh_token: this.token.refreshToken
              },
              responseJson
            )
          );
          this._storeTokenData(this.token);
        } else {
          // Go back to the no token path if refresh failed
          // This will most likely result in an auth redirect
          this.token = null;
          await this.refreshAccessToken();
      }
      }
    } catch (error) {
      const tokenFromStorage = this._getTokenDataFromStorage();
      if (tokenFromStorage) {
        this.token = this.client.createToken(tokenFromStorage as any);
        this.token.expiresIn(new Date(tokenFromStorage.expires));
        if (this._isTokenExpiring(options)) {
          /**
           * Remove token from storage. It will be added back on successful token refresh
           * The other possibility is that there is an error refreshing the token and this
           * token should be thrown out until the user is authenticated again
           */
          this._removeTokenDataFromStorage();
          await this.refreshAccessToken({ expiringRefresh: true });
        }
      } else {
        console.warn(error, 'Error refreshing access token - redirecting');
        window.location.href = this.client.code.getUri();
      }
    }
  }

  public async startAutomaticTokenRefresh(options: LOAuth.RefreshOptions) {
    if (!this.token) {
      // Initiate initial auth token exchange
      await this.refreshAccessToken(options);
    }

    options = options || {};
    if (!this.refreshInterval) {
      this.refreshInterval = window.setInterval(async () => {
        try {
          if (this._isTokenExpiring(options)) {
            await this.refreshAccessToken({ expiringRefresh: true });
          }
        } catch (error) {
          console.warn(error, 'Error in automatic token refresh');
        }
      }, options.interval || AUTH_TOKEN_INTERVAL);
    }
  }

  public async stopAutomaticTokenRefresh() {
    if (this.refreshInterval) {
      window.clearInterval(this.refreshInterval);
    }
  }

  public getAccessToken() {
    return this.token && this.token.accessToken;
  }

  public async sign(options: LOAuth.SignOptions) {
    if (!this.token) {
      throw new Error(
        'Cannot sign request before receiving access token - wait for refreshAccessToken'
      );
    }
    return this.token.sign(options);
  }

  public async logout() {
    await this.stopAutomaticTokenRefresh();
    this._removeTokenDataFromStorage();
    const qs = queryString.stringify({
      client_id: this.clientOptions.clientId,
      logout_uri: this.clientOptions.logoutRedirectUri
    });
    window.location.href = `${this.clientOptions.logoutUri}?${qs}`;
  }
}

declare namespace LOAuth {
  export interface Config {
    clientId: string;
    authorizationUri: string;
    accessTokenUri: string;
    redirectUri: string;
    logoutUri: string;
    logoutRedirectUri: string;
    scopes: string[];
    storageKey?: string;
    storage?: Storage;

    // An object containing application state
    appState?: any;
  }

  export interface Token extends ClientOAuth2.Token {
    expires?: Date;
    data: any;
  }

  export interface RefreshOptions extends ClientOAuth2.Options {
    expiringRefresh?: boolean;
    refreshWindow?: number;
    interval?: number;
  }

  export interface SignOptions extends ClientOAuth2.RequestObject {}

  export interface Storage {
    getItem(key: string): string;
    setItem(key: string, value: string): void;
    removeItem(key: string): void;
  }

  export interface TokenData {
    access_token: string;
    expires_in: number;
    id_token: string;
    refresh_token: string;
    token_type: string;
    expires?: number;
  }
}

export default LOAuth;
